Make optimisations to resolver implementation

yamadapc opened the pull request:
Replaced by https://github.com/parcel-bundler/parcel/pull/9903

The main change in this diff is removing an extra `is_dir` / `is_file` call which I had added.

We ran benchmarks replaying all imports of a large application against multiple resolver implementations. In practice, the previous code had around 10% worse performance than the shared structures code. The difference in performance goes away by removing those extra `is_dir` and `is_file` calls. There is no difference achieved from preventing copies.

I'll attach performance profiles captured on macOS.

After the fixes in this commit, the new resolver crate out-performs the older one by 26-27%.

That is due to a change on top of the FS calls revert, which I've found through profiling. Hashes have been replaced for both implementations' filesystem canonicalize cache, which is the hottest by a faster hashing function.

The worse culprit of CPU time at the moment is hashing file-paths on this benchmark. This might be due to how the file-paths are passed into the benchmark, but we're constantly re-hashing file-path components to store in these cache hash-maps.

The total time of resolving all the application level imports on multiple threads (~500k) on this application is:

* 2.13 seconds for M1 Pro (new branch)
* 2.88 seconds for M1 Pro (old branch)

This is consistent across runs and based on an average of 50 executions over this entire list.

On a large Linux EC2 instance the results are similar, but resolution is even faster.

After path hashing, the next hottest path are the file-system calls.

DashMap does not appear to have a huge impact on performance on its own, and is showing-up on crash reports. I'm adding it back hoping it's not the culprit.

- - -

Note, this is relatively incredibly fast and therefore not a focus for performance optimisation.

We're still experiencing crashes from this area, which we'll investigate further.

yamadapc committed:
Make optimisations to resolver implementation

The main change in this diff is removing an extra `is_dir` / `is_file` call
which I had added.

We ran benchmarks replaying all imports of a large application against multiple
resolver implementations. In practice, the previous code had around 10% worse
performance than the shared structures code. The difference in performance goes
away by removing those extra `is_dir` and `is_file` calls. There is no
difference achieved from preventing copies.

I'll attach performance profiles captured on macOS.

After the fixes in this commit, the new resolver crate out-performs the older
one by 26-27%.

That is due to a change on top of the FS calls revert, which I've found through
profiling. Hashes have been replaced for both implementations' filesystem
canonicalize cache, which is the hottest by a faster hashing function.

The worse culprit of CPU time at the moment is hashing file-paths on this
benchmark. This might be due to how the file-paths are passed into the
benchmark, but we're constantly re-hashing file-path components to store in
these cache hash-maps.

The total time of resolving all the application level imports on multiple
threads (~500k) on this application is:

* 2.13 seconds for M1 Pro (new branch)
* 2.88 seconds for M1 Pro (old branch)

This is consistent across runs and based on an average of 50 executions over
this entire list.

On a large Linux EC2 instance the results are similar, but resolution is even
faster.

After path hashing, the next hottest path are the file-system calls.

DashMap does not appear to have a huge impact on performance on its own, and is
showing-up on crash reports. I'm adding it back hoping it's not the culprit.

MonicaOlejniczak commented on packages/utils/node-resolver-rs/src/cache.rs:
```diff
@@ -86,15 +89,13 @@ impl JsonError {
-      is_file_cache: DashMap::new(),
-      is_dir_cache: DashMap::new(),
-      realpath_cache: DashMap::new(),
+      packages: DashMap::with_hasher(gxhash::GxBuildHasher::default()),
```
Last I checked this won't compile on musl and some other platforms

MonicaOlejniczak commented on crates/parcel_filesystem/Cargo.toml:
```diff
@@ -7,6 +7,8 @@ description = "FileSystem wrapper trait for use in Parcel codebase."
 [dependencies]
 dashmap = "5.5.3"
 mockall = "0.12.1"
+gxhash = "3.4.1"
```
Last I checked this won't compile on musl and some other platforms

You also need to enable aes

yamadapc replied:
Changing to xxhash doesn't change the results very much so I think I'll replace with that.

devongovett commented on packages/utils/node-resolver-rs/src/cache.rs:
```diff
@@ -22,11 +22,14 @@ pub struct Cache {
+  // In particular just the is_dir_cache spends around 8% of the time on a large project resolution
+  // hashing paths. Instead of using a hashmap we should try a trie here.
+  is_dir_cache: DashMap<PathBuf, bool, gxhash::GxBuildHasher>,
+  is_file_cache: DashMap<PathBuf, bool, gxhash::GxBuildHasher>,
```
Something I found before but never got around to committing was that making the keys `OsString` instead of `PathBuf` made the hashing significantly faster. I think that should be equivalent as long as the path separators are normalized (but even then it shouldn't matter for correctness, would just result in multiple entries for the same file).

yamadapc replied:
Most of the hashing time is spent in the function body of the `Path` hash function (rather than the hasher).

So I think it might be faster to just use strings, but the hashes will be different if any path key isn't normalized yes.

yamadapc committed:
Use xxhash instead of gxhash

This did not change the results of the previous commit 5682caa3ed445f8faada1a1675565e22bcf157ca significantly.

yamadapc committed:
Fix compilation failures

yamadapc committed:
Fix compiling tests

yamadapc committed:
Fix unit-test compilation

yamadapc commented:
I suppose the REPL deployment is failing due to this being on my own github fork rather than this repository (which was an accident.

I will re-open this PR.
